fdbrpc/include/fdbrpc/Msgpack.h

```diff
@@ -0,0 +1,157 @@
+/*
+ * Msgpack.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef FDBRPC_MSGPACK_H
+#define FDBRPC_MSGPACK_H
+#include <limits>
+#pragma once
+
+#include <memory>
+#include <algorithm>
+#include "flow/Trace.h"
+#include "flow/Error.h"
+#include "flow/network.h"
+
+struct MsgpackBuffer {
+	std::unique_ptr<uint8_t[]> buffer;
+	// Amount of data in buffer (bytes).
+	std::size_t data_size;
+	// Size of buffer (bytes).
+	std::size_t buffer_size;
+
+	void write_byte(uint8_t byte) { write_bytes(&byte, 1); }
+
+	// This assumes that pos <= data_size
+	void edit_byte(uint8_t byte, size_t pos) { buffer[pos] = byte; }
+
+	void write_bytes(const uint8_t* buf, std::size_t n) {
+		resize(n);
+		std::copy(buf, buf + n, buffer.get() + data_size);
+		data_size += n;
+	}
+
+	void resize(std::size_t n) {
+		if (data_size + n <= buffer_size) {
+			return;
+		}
+
+		std::size_t size = buffer_size;
+		while (size < data_size + n) {
+			size *= 2;
+		}
+
+		TraceEvent(SevInfo, "MsgpackResizedBuffer").detail("OldSize", buffer_size).detail("NewSize", size);
+		auto new_buffer = std::make_unique<uint8_t[]>(size);
+		std::copy(buffer.get(), buffer.get() + data_size, new_buffer.get());
+		buffer = std::move(new_buffer);
+		buffer_size = size;
+	}
+
+	void reset() { data_size = 0; }
+};
+
+inline void serialize_bool(bool val, MsgpackBuffer& buf) {
+	if (val) {
+		buf.write_byte(0xc3);
+	} else {
+		buf.write_byte(0xc2);
+	}
+}
+
+// Writes the given value in big-endian format to the request. Sets the
+// first byte to msgpack_type.
+template <typename T>
+inline void serialize_value(const T& val, MsgpackBuffer& buf, uint8_t msgpack_type) {
```

Comment:
Just a thought for the future: these functions could be named `serialize` and use the parameters to allow the compiler to deduce which function is called, instead of explicitly calling out the types in each serialize function name. This doesn't need to be changed now since you are just refactoring existing code, but a thought for the future.

Reply:
That would be cool. I thought about it, but I wasn't sure how to make it map `T` to `msgpack_type` at compile time.